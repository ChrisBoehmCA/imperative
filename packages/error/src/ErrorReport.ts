/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { IErrorReportParms } from "./../../error";
import { isNullOrUndefined } from "util";
import { ImperativeError } from "./ImperativeError";
import { IErrorSavedSettings } from "./doc/IErrorSavedSettings";

/**
 * Class to make use of node-report as an optional dependency.  This is optional
 * because it is a native module with native dependencies.
 * @export
 * @class ErrorReport
 */
export class ErrorReport {

    /**
     * Message for uninstalled node-report
     * @static
     * @memberof ErrorReport
     */
    public static readonly NOT_INSTALLED_MESSAGE = "Node-report is not installed";

    /**
     * Return a report to the called to be displayed or logged in the format of the
     * caller's choosing.
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @returns {string} - returns string report
     * @memberof ErrorReport
     */
    public static obtain(parms?: IErrorReportParms): string {
        const settings = ErrorReport.setEnvironment(parms);
        let report: string;
        try {
            const nodeReport = require("node-report");
            report = nodeReport.getReport();
        } catch (error) {
            report = ErrorReport.NOT_INSTALLED_MESSAGE;
        } finally {
            ErrorReport.restoreEnvironment(settings);
        }
        return report;
    }

    /**
     * Trigger a report to be generated as a text document.
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @memberof ErrorReport
     */
    public static trigger(parms?: IErrorReportParms) {
        const settings = ErrorReport.setEnvironment(parms);
        try {
            const nodeReport = require("node-report");
            nodeReport.triggerReport();
        } catch (error) {
            // return ErrorReport.NOT_INSTALLED_MESSAGE;
        } finally {
            ErrorReport.restoreEnvironment(settings);
        }
    }

    /**
     * Set environment variables for node report prior to using it.
     * @private
     * @static
     * @param {IErrorReportParms} parms - parms object to configure the report (see IErrorReportParms)
     * @memberof ErrorReport
     */
    private static setEnvironment(parms?: IErrorReportParms) {
        const settings: IErrorSavedSettings = {};
        settings.NODEREPORT_VERBOSE = "no";
        if (!isNullOrUndefined(parms)) {
            if (!isNullOrUndefined(parms.verbose)) {
                process.env.NODEREPORT_VERBOSE = parms.verbose;
            }
        }
        return settings;
    }

    /**
     * Restore environmental settings
     * @private
     * @static
     * @param {IErrorSavedSettings} settings - object to contain restoration values for env variable
     * @memberof ErrorReport
     */
    private static restoreEnvironment(settings: IErrorSavedSettings) {
        process.env.NODEREPORT_VERBOSE = settings.NODEREPORT_VERBOSE;
    }
}
