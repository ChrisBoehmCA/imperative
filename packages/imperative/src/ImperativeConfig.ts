/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright Contributors to the Zowe Project.
*
*/

import { Constants } from "../../constants";
import { CommandPreparer, ICommandDefinition, ICommandProfileTypeConfiguration } from "../../cmd";
import { CompleteProfilesGroupBuilder } from "./profiles/builders/CompleteProfilesGroupBuilder";
import { DefinitionTreeResolver } from "./DefinitionTreeResolver";
import { Logger } from "../../logger";
import { dirname, join } from "path";
import { IImperativeConfig } from "./doc/IImperativeConfig";
import { ImperativeError } from "../../error";
import { EnvironmentalVariableSettings } from "./env/EnvironmentalVariableSettings";

/**
 * This class is used to contain all configuration being set by Imperative.
 * It is a singleton and should be accessed via ImperativeConfig.instance.
 */
export class ImperativeConfig {
    /**
     * This is the variable that stores the specific instance of Imperative Config.
     * Defined as static so that it can be accessed from anywhere.
     *
     * @private
     * @type {ImperativeConfig}
     */
    private static mInstance: ImperativeConfig = null;


    /**
     * This variable is used to contain an instance of the Logger object.
     *
     * @private
     * @type {Logger}
     */
    private console: Logger = Logger.getImperativeLogger();

    /**
     * This parameter is used as the container of all loaded configuration for
     * Imperative.
     *
     * @private
     * @type {IImperativeConfig}
     */
    private mLoadedConfig: IImperativeConfig = null;

    /**
     * This parameter is used to contain the caller location of imperative configuration file.
     *
     * @private
     * @type {string}
     */
    private mCallerLocation: string = null;

    /**
     * This is the package name of the host application. It will only be set once accessed to
     * lessen loads to the host package.json.
     */
    private mHostPackageName: string;

    /**
     * This is the name of our imperative package. It will only be set once accessed to
     * lessen loads to the imperative package.json.
     *
     * It isn't hardcoded so that the name of our package can change without affecting
     * modules dependent on it.
     */
    private mImperativePackageName: string;

    /**
     * Gets a single instance of the PluginIssues. On the first call of
     * ImperativeConfig.instance, a new Plugin Issues object is initialized and returned.
     * Every subsequent call will use the one that was first created.
     *
     * @returns {ImperativeConfig} The newly initialized PMF object.
     */
    public static get instance(): ImperativeConfig {
        if (this.mInstance == null) {
            this.mInstance = new ImperativeConfig();
        }

        return this.mInstance;
    }

    /**
     * Set the caller location.
     * @param {string} location new location to be updated with
     */
    public set callerLocation(location: string) {
        this.mCallerLocation = location;
    }

    /**
     * Return file location of imperative configuration file.
     * @returns {string} - location of configuration file
     */
    public get callerLocation(): string {
        return this.mCallerLocation;
    }

    /**
     * Set the loaded config data.
     * @param {IImperativeConfig} config to be set.
     */
    public set loadedConfig(config: IImperativeConfig) {
        this.mLoadedConfig = config;
    }

    /**
     * Retrieve the loaded config (if init has
     * @returns {IImperativeConfig} - the config that has been loaded, if any
     */
    public get loadedConfig(): IImperativeConfig {
        return this.mLoadedConfig;
    }

    /**
     * Retrieve the host package name from which imperative was called.
     */
    public get hostPackageName(): string {
        if (!this.mHostPackageName) {
            this.mHostPackageName = this.callerPackageJson.name;
        }

        return this.mHostPackageName;
    }

    /**
     * Retrieve the package name of the imperative application.
     */
    public get imperativePackageName(): string {
        if (!this.mImperativePackageName) {
            this.mImperativePackageName = require(join(__dirname, "../../../package.json")).name;
        }

        return this.mImperativePackageName;
    }

    /**
     * Add a new command group by inserting it to the definitions list of the loaded config.
     * @param {ICommandDefinition} cmdDefToAdd - command definition group to to be added.
     */
    public addCmdGrpToLoadedConfig(cmdDefToAdd: ICommandDefinition): void {
        if (this.loadedConfig != null) {
            if (this.loadedConfig.definitions == null) {
                this.loadedConfig.definitions = [];
            }
            const defIndex = this.loadedConfig.definitions.indexOf(cmdDefToAdd);
            if (defIndex > -1) {
                this.loadedConfig.definitions.splice(defIndex, 1);
            }
            this.console.debug("Adding definition = '" + cmdDefToAdd.name + "'");
            this.loadedConfig.definitions.push(cmdDefToAdd);
        }
    }

    /**
     * Add a new set of profiles by inserting them into the profiles of the loaded config.
     * @param {ICommandProfileTypeConfiguration[]} profiles
     *    Array of profiles to be added.
     */
    public addProfiles(profiles: ICommandProfileTypeConfiguration[]): void {
        if (this.loadedConfig) {
            if (!this.loadedConfig.profiles) {
                this.loadedConfig.profiles = [];
            }
            for (const profileToAdd of profiles) {
                /* We expect to not find (ie, undefined result) an exiting profile
                 * with the same type value as the profile that we want to add.
                 */
                const existingProfile: ICommandProfileTypeConfiguration =
                    this.loadedConfig.profiles.find((profileToTest) => {
                        return profileToTest.type === profileToAdd.type;
                    });
                if (existingProfile) {
                    this.console.error("addProfiles: The profile of type '" + profileToAdd.type +
                        "' already exists. It will not be added.");
                    continue;
                }
                this.console.debug("addProfiles: Adding " + profileToAdd.type + " profile");
                this.loadedConfig.profiles.push(profileToAdd);
            } // end for
        } // end if loadedConfig not null
    }

    /**
     * Parses the package.json file and searches for the symlink name used under "bin".
     * @returns {string} - return bin symlink name if present, otherwise null
     */
    public findPackageBinName(): string {
        const pkg = this.callerPackageJson;
        if (typeof pkg.bin === "string") {
            return pkg.name;
        } else if (typeof pkg.bin === "object") {
            return Object.keys(pkg.bin).pop();
        }
        return null;
    }

    /**
     * Return the cli Home path.
     * @return {string} path to cli Home.
     */
    public get cliHome(): string {
        const settings = EnvironmentalVariableSettings.read(this.loadedConfig.envVariablePrefix || this.loadedConfig.name);
        if (settings.cliHome.value != null) {
            return settings.cliHome.value;
        }
        return this.loadedConfig.defaultHome;
    }

    /**
     * Return profile Directory.
     * @return {string} profile directory.
     */
    public get profileDir(): string {
        return this.loadedConfig.defaultHome + Constants.PROFILES_DIR + "/";
    }

    /**
     * Get imperative's host CLI command tree with all module globs resolved.
     *
     * @return {ICommandDefinition} The resolved command tree
     */
    public get resolvedCmdTree(): ICommandDefinition {
        const config = this.loadedConfig;
        return DefinitionTreeResolver.resolve(config.rootCommandDescription || "",
            config.productDisplayName,
            dirname(this.callerLocation),
            this.console,
            config.definitions, config.commandModuleGlobs
        );
    }

    /**
     * Get imperative's host CLI command tree after final preparation.
     *
     * @param {IImperativeConfig} resolvedCmdTree - The imperative command tree
     *        returned by ImperativeConfig.resolvedCmdTree()
     */
    public getPreparedCmdTree(resolvedCmdTree: ICommandDefinition): ICommandDefinition {
        let preparedCmdTree = this.addAutoGeneratedCommands(resolvedCmdTree);
        preparedCmdTree = CommandPreparer.prepare(preparedCmdTree);
        return preparedCmdTree;
    }

    /**
     * Return package.json of the imperative user
     * @returns {any} - package.json file of caller
     */
    public get callerPackageJson(): any {
        return this.getCallerFile("package.json");
    }

    /**
     * Require a file from a project using imperative accounting for imperative being contained
     * separately from the current implementers directory.
     * @param {string} file - the file to require from project using imperative
     */
    public getCallerFile(file: string): any {
        // try to locate the file using find-up first
        let findupErr: Error;
        try {
            const filePath = require("find-up").sync(file, {cwd: ImperativeConfig.instance.callerLocation});
            return require(filePath);
        } catch (e) {
            // couldn't locate using find-up, try to require directly
            findupErr = e;
        }
        // if we couldn't find the file path through find-up, try requiring the string directly
        try {
            return require(file);
        } catch (e) {
            e.message = "Could not locate the specified module through requiring directly, nor through " +
                "searching the directories above " + this.callerLocation +
                ". 'require()' error message: " + e.message +
                " \n 'find-up' (directory search) error message:" + findupErr.message;
            throw new ImperativeError({msg: e.message});
        }
    }

    /**
     * Append any auto generated commands to the root command document depending on configuration.
     * @param {ICommandDefinition} rootCommand - the root command as built so far
     * @returns {ICommandDefinition} - the root command with any auto generated commands appended
     */
    private addAutoGeneratedCommands(rootCommand: ICommandDefinition): ICommandDefinition {
        const loadedConfig: IImperativeConfig = this.loadedConfig;
        if ((loadedConfig.autoGenerateProfileCommands == null || loadedConfig.autoGenerateProfileCommands) &&
            loadedConfig.profiles != null &&
            loadedConfig.profiles.length > 0) {
            rootCommand.children.push(CompleteProfilesGroupBuilder.getProfileGroup(loadedConfig.profiles, this.console));
        }
        return rootCommand;
    }

}
