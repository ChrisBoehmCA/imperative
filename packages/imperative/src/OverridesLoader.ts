/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*/

import {IImperativeOverrides} from "./doc/IImperativeOverrides";
import { CredentialManagerFactory, DefaultCredentialManager } from "../../security";
import {IImperativeConfig} from "./doc/IImperativeConfig";
import { isAbsolute, resolve } from "path";

/**
 * Imperative-internal class to load overrides
 * You should not need to call this from your CLI.
 */
export class OverridesLoader {
  /**
   * Apply overrides to all applicable facilities and use our defaults where
   * an override is not provided.
   *
   * @param {IImperativeConfig} config - the current {@link Imperative#loadedConfig}
   */
  public static async load(
    config: IImperativeConfig
  ): Promise<void> {
    // Initialize the Credential Manager
    await this.loadCredentialManager(config);
  }

  /**
   * Initialize the Credential Manager using the supplied override when provided.
   *
   * @param {IImperativeConfig} config - the current {@link Imperative#loadedConfig}
   */
  private static async loadCredentialManager(
    config: IImperativeConfig
  ): Promise<void> {
    const overrides: IImperativeOverrides = config.overrides;

    // If the credential manager wasn't set, then we use the DefaultCredentialManager
    if (overrides.CredentialManager == null) {
      overrides.CredentialManager = DefaultCredentialManager;
    }
    // If the credential manager is type string and not absolute, we will convert it to an absolute path
    // relative to the process entry file location.
    else if (typeof overrides.CredentialManager === "string" && !isAbsolute(overrides.CredentialManager)) {
      overrides.CredentialManager = resolve(process.mainModule.filename, "../", overrides.CredentialManager);
    }

    await CredentialManagerFactory.initialize(overrides.CredentialManager, config.name);
  }
}
