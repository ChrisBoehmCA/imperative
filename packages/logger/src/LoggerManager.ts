/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import {IQueuedMessage} from "./doc/IQueuedMessage";
import { Console } from "../../console";

/**
 * LoggerManager is a singleton class used to contain logger information.
 */

export class LoggerManager {
    private static readonly DEFAULT_MAX_QUEUE_SIZE = 10000;
    private static mInstance: LoggerManager = null;

    public static get instance(): LoggerManager {
        if (this.mInstance == null) {
            this.mInstance = new LoggerManager();
        }

        return this.mInstance;
    }

    private mIsLoggerInit: boolean = false;
    private mLogInMemory: boolean = false;
    private mMaxQueueSize: number;
    private console: Console;
    private mQueuedMessages: IQueuedMessage[] = [];

    constructor() {
        this.console = new Console();
        this.mMaxQueueSize = LoggerManager.DEFAULT_MAX_QUEUE_SIZE;
    }

    public get isLoggerInit(): boolean {
        return this.mIsLoggerInit;
    }

    public set isLoggerInit(status: boolean) {
        this.mIsLoggerInit = status;
    }

    public get logInMemory(): boolean {
        return this.mLogInMemory;
    }

    public set logInMemory(status: boolean) {
        this.mLogInMemory = status;
    }

    public get maxQueueSize(): number {
        return this.mMaxQueueSize;
    }

    public set maxQueueSize(size: number){
        this.mMaxQueueSize = size;
    }

    /**
     * This function is responsible for gathering all of the input parameters and
     * store them in the message queue array.
     * New messages are to be stored at the top of the array instead of the bottom.
     * This allow easy removing message from array while looping the array.
     * @param category - logger category
     * @param method - log method
     * @param message - log message
     */
    public queueMessage(category: string, method: string, message: string){
        if (this.logInMemory) {
            this.mQueuedMessages.unshift({
                category,
                method,
                message
            });

            if (this.mQueuedMessages.length > this.maxQueueSize) {
                this.mQueuedMessages.pop();
            }
        } else {
            this.console.info(message);
        }
    }

    public get QueuedMessages(): IQueuedMessage[] {
        return this.mQueuedMessages;
    }

}
