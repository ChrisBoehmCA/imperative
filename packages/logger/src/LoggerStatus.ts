/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import {IQueuedMessage} from "./doc/IQueuedMessage";

/**
 * LoggerStatus is a singleton class used to contain logger information.
 */

export class LoggerStatus {

    private static mInstance: LoggerStatus = null;

    public static get instance(): LoggerStatus {
        if (this.mInstance == null) {
            this.mInstance = new LoggerStatus();
        }

        return this.mInstance;
    }


    private mIsLoggerInit: boolean = false;
    private mQueuedMessages: IQueuedMessage[] = [];

    public get isLoggerInit(): boolean {
        return this.mIsLoggerInit;
    }

    public set isLoggerInit(status: boolean) {
        this.mIsLoggerInit = status;
    }

    /**
     * This function is responsible for gathering all of the input parameters and
     * store them in the message queue array.
     * New messages are to be stored at the top of the array instead of the bottom.
     * This allow easy removing message from array while looping the array.
     * @param category - logger category
     * @param method - log method
     * @param message - log message
     */
    public queueMessage(category: string, method: string, message: string){
        this.mQueuedMessages.unshift({
            category,
            method,
            message
        });
    }

    public get QueuedMessages(): IQueuedMessage[] {
        return this.mQueuedMessages;
    }


}
