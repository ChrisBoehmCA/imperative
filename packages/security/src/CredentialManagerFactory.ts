/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import { AbstractCredentialManager } from "./abstract/AbstractCredentialManager";
import { ImperativeError } from "../../error";
import { IImperativeOverrides } from "../../imperative/src/doc/IImperativeOverrides";

/**
 * This is a wrapper class that controls access to the credential manager used within
 * the Imperative framework. All calls to the credential manager done by Imperative
 * must go through this class for security reasons.
 */
export class CredentialManagerFactory {
    /**
     * Initialize the credential manager, then lock the door and throw away the
     * key. This method can only be called once and should be called by {@link Imperative.init}
     * immediately after the CLI configuration has been loaded.
     *
     * This is where any Credential Manager your cli provides will be initialized. First
     * Imperative will instantiate your manager (or the {@link DefaultCredentialManager} if none was provided to
     * {@link Imperative.init}) and will then call your class's initialize method.
     *
     * ### Dynamic Import of Module
     *
     * This method will perform a dynamic import of your {@link IImperativeOverrides.CredentialManager} module when the
     * Manager parameter is passed as a string. If anything goes wrong during this import or if the module that was exported
     * doesn't extend the {@link AbstractCredentialManager}, this method will throw an error.
     *
     * @see {@link IImperativeOverrides.CredentialManager}
     *
     *
     * ### Immutable Class Creation
     *
     * After this method is complete, the instantiated credential manager will no longer allow changes
     * to it's direct variable assignments. This means that even your class can only change the values of it's direct
     * properties in the constructor and the initialize method. However, this does not prevent you from changing values
     * of properties of one of your classes objects.
     *
     * For example, after initialization, your class can not do something like this: `this.someProp = 5`. This will result
     * in a JavaScript "Cannot assign to read only property" exception because your class is immutable.
     * You still will be able to do stuff like this if someProp was already an object: `this.someProp.someValue = 5`. This
     * occurs because while Imperative marks your class as immutable (using Object.freeze) the underlying `this.someProp`
     * object is still mutable.
     *
     * @see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/freeze
     *
     * @param {IImperativeOverrides.CredentialManager} Manager - A class that extends {@link AbstractCredentialManager} that will
     *                                                  be instantiated and used as the actual credential manager. If a string is
     *                                                  passed, we will attempt to load the module specified in the string as a
     *                                                  class that extends the __AbstractCredentialManager__. If the class imported
     *                                                  doesn't extend the abstract class, we will throw an error.
     *
     * @param {string} cliName - The cli name to be used in the security manager. This really is only required for the
     *                           default manager so that the service name for keytar can be the cli name. It may be useful
     *                           for a custom implementation, so it will be passed every time and it is up to the
     *                           implementer to choose to use it.
     *
     * @throws {@link ImperativeError} When it has been detected that this method has been called before.
     *         It is important that this method only executes once.
     *
     * @throws {@link ImperativeError} When the file specified by the Manager string references a module that
     *         does not extend {@link AbstractCredentialManager}.
     *
     * @throws {@link ImperativeError} When the file specified by the Manager string does not exist or does not
     *         export a class as part of the module.exports clause.
     */
    public static async initialize(Manager: IImperativeOverrides["CredentialManager"], cliName: string): Promise<void> {
        if (this.mManager != null) {
            // Something tried to change the already existing credential manager, we should stop this.
            throw new ImperativeError({
                msg: "A call to CredentialManagerFactory.initialize has already been made! This method can only be called once",
            });
        }

        // @TODO TEST LOGIC
        try {
            let manager: any;

            // Dynamically determine which manager to load.
            if (typeof Manager === "string") {
                // In the case of a string, we make the assumption that it is pointing to the absolute file path of something
                // that exports a manager class. So we'll load that class and initialize it with the same constructor parameters
                // that we would do with an actual constructor parameter.
                const LoadedManager = await import(Manager);
                manager = new LoadedManager(cliName);
            } else {
                manager = new Manager(cliName);
            }

            // After constructing the object, we will ensure that the thing loaded is indeed an
            // instance of an abstract credential manager. Since we cannot assume that our internal
            // load of a plugin provided a correct object to the function :/
            if (manager instanceof AbstractCredentialManager) {
                this.mManager = manager;
            } else {
                const message = (typeof Manager === "string") ?
                    `The manager provided at ${Manager} does not extend AbstractCredentialManager properly!` :
                    "A bad object was provided to the CredentialManagerFactory.initialize() method. This could be " +
                    "due to a bad plugin.";

                throw new ImperativeError({
                    msg: message
                });
            }

            if (this.mManager.initialize) {
                await this.mManager.initialize();
            }
        } catch (error) {
            // Perform dynamic requires when an error happens
            const { InvalidCredentialManager } = await import("./InvalidCredentialManager");
            const { Logger } = await import("../../logger");
            const appSettings  = (await import("../../settings")).AppSettings.instance;

            // A value not equal to false indicates that the setting was overridden by a plugin
            // so we should not have a hard crash.
            if (appSettings.settings.overrides.CredentialManager !== false) {
                const logError = "Failed to override the credential manager with one provided by \"" +
                    appSettings.settings.overrides.CredentialManager +
                    "\"";

                // Be sure to log the messages both to the console and to a file
                // so that support can also see these messages.
                Logger.getImperativeLogger().error(logError);
                Logger.getConsoleLogger().error(logError);

                Logger.getImperativeLogger().error(error.toString());
                Logger.getConsoleLogger().error(error.toString());

                this.mManager = new InvalidCredentialManager(cliName, error);
            } else {
                this.mManager = undefined;

                // The crash was caused by a bad override provided by a base cli
                // so this should be thrown up as a hard crash
                throw error;
            }
        }

        // Freeze both the wrapper class and the credential manager we just created
        // to prevent anyone from making changes to this class after the first
        // initialization. This plugs up a security hole so that a plugin can never
        // trash the security manager created on init.
        Object.freeze(this);
        Object.freeze(this.mManager);
    }

    /**
     * Static singleton instance of an instantiated {@link AbstractCredentialManager}
     *
     * @private
     */
    private static mManager: AbstractCredentialManager;

    /**
     * @returns {AbstractCredentialManager} - The credential manager that Imperative should use to
     *   retrieve user credentials.
     *
     * @throws {ImperativeError} - When the Credential Manager has not been initialized yet.
     */
    public static get manager(): AbstractCredentialManager {
        if (this.mManager == null) {
            throw new ImperativeError({
                msg: "Credential Manager not yet initialized! CredentialManagerFactory.initialize must " +
                    "be called prior to CredentialManagerFactory.mananger"
            });
        }

        return this.mManager;
    }
}

// This also prevents someone being able to hijack the exported object. Essentially all of these Object.freeze
// calls make this class the programming equivalent of Fort Knox. Hard to get in but if you do get in, then
// there are lots of valuables
Object.freeze(exports);
