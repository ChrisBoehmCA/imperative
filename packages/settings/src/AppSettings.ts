/*
* This program and the accompanying materials are made available under the terms of the *
* Eclipse Public License v2.0 which accompanies this distribution, and is available at *
* https://www.eclipse.org/legal/epl-v20.html                                      *
*                                                                                 *
* SPDX-License-Identifier: EPL-2.0                                                *
*                                                                                 *
* Copyright Contributors to the Zowe Project.                                     *
*                                                                                 *
*/

import {readFileSync, writeFile} from "jsonfile";
import {existsSync} from "fs";
import {ISettingsFile} from "./doc/ISettingsFile";
import {Logger} from "../../logger";

/**
 * A recovery function to handle when a settings file is missing on an initialization function.
 *
 * @param settingsFile The file that could not be found
 * @param defaultSettings The default settings provided by {@link AppSettings}
 *
 * @returns The settings content to be merged into the defaults.
 */
export type FileRecovery = (settingsFile: string, defaultSettings: ISettingsFile) => ISettingsFile;

/**
 * This class represents settings for an Imperative CLI application that can be configured
 * by an end user by modifying a settings file. @TODO expand on this doc
 */
export class AppSettings {
    /**
     *
     * @param settingsFile The settings file to load from.
     * @param missingFileRecovery A recovery function when the settings file isn't found
     *
     * @throws {@link SettingsAlreadyInitialized} When the settings singleton has previously been initialized.
     */
    public static initialize(settingsFile: string, missingFileRecovery ?: FileRecovery): AppSettings {
        if (AppSettings.mInstance) {
            // Throw an error imported at runtime so that we minimize file that get included
            // on startup.
            const { SettingsAlreadyInitialized } = require("./errors/index");
            throw new SettingsAlreadyInitialized();
        }

        AppSettings.mInstance = new AppSettings(settingsFile, missingFileRecovery);
        return AppSettings.mInstance;
    }


    /**
     * This is an internal reference to the static settings instance.
     */
    private static mInstance: AppSettings;

    /**
     * Get the singleton instance of the app settings object that was initialized
     * within the {@link AppSettings.initialize} function.
     *
     * @returns A singleton AppSettings object
     *
     * @throws {@link SettingsNotInitialized} When the settings singleton has not been initialized.
     */
    public static get instance(): AppSettings {
        if (AppSettings.mInstance == null) {
            // Throw an error imported at runtime so that we minimize file that get included
            // on startup.
            const { SettingsNotInitialized } = require("./errors/index");
            throw new SettingsNotInitialized();
        }

        return AppSettings.mInstance;
    }

    /**
     * Internal reference to the overrides settings. The defaults should
     * all be false, indicating that there are no overrides to be done.
     */
    public readonly settings: ISettingsFile;

    /**
     * Constructs a new settings object
     *
     * @param settingsFile  The full path to a settings file to load.
     * @param missingFileRecovery A recovery function for when the settings file didn't exist
     */
    constructor(settingsFile: string, missingFileRecovery ?: FileRecovery) {
        let settings: ISettingsFile;
        const defaultSettings: ISettingsFile = {
            overrides: {
                CredentialManager: false
            }
        };

        try {
            // Try to load the file immediately, if it fails we will then
            // try to recover
            Logger.getImperativeLogger().trace(`Attempting to load settings file: ${settingsFile}`);
            settings = readFileSync(settingsFile);
            Logger.getImperativeLogger().trace("Settings were loaded");
        } catch (up) {
            if (missingFileRecovery && !existsSync(settingsFile)) {
                Logger.getImperativeLogger().trace("Executing missing file recovery.");
                settings = missingFileRecovery(settingsFile, defaultSettings);
            } else {
                Logger.getImperativeLogger().error("Unable to recover from load failure");
                Logger.getImperativeLogger().error(up.toString());
                // Throw up if there is no recovery function or there was a recovery
                // function but the file already existed. (Indicates there was a bigger
                // issue at play)
                throw up;
            }
        }

        // Merge objects loaded recursively
        const mergeObjects = require("merge-objects");
        this.settings = mergeObjects(defaultSettings, settings);

        Logger.getImperativeLogger().trace("Loaded Settings:");
        Logger.getImperativeLogger().trace(this.settings as any); // This works because someone does the object translation

        writeFile(settingsFile, this.settings, {
            spaces: 2
        }, (err) => {
            if (err) {
                Logger.getImperativeLogger().error("Unable to save settings");
                Logger.getImperativeLogger().error(err.toString());
            }
        });
    }
}
