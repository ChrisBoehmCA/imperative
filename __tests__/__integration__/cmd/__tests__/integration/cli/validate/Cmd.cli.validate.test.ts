import {ITestEnvironment} from "../../../../../../__src__/environment/doc/response/ITestEnvironment";
import {SetupTestEnvironment} from "../../../../../../__src__/environment/SetupTestEnvironment";
import {runCliScript} from "../../../../../../src/TestUtil";

let TEST_ENVIRONMENT: ITestEnvironment;

describe("cmd-cli", () => {

    // Create the unique test environment
    beforeAll(async () => {
        TEST_ENVIRONMENT = await SetupTestEnvironment.createTestEnv({
            cliHomeEnvVar: "CMD_CLI_CLI_HOME",
            testName: "cmd_validate"
        });
    });

    it("should display proper messages when missing positional parameter", async () => {
        const response = runCliScript(__dirname + "/__scripts__/invalid_positional1.sh", TEST_ENVIRONMENT.workingDir);
        expect(response.status).toBe(1);
        expect(response.stderr.toString()).toMatchSnapshot();
        expect(response.stdout.toString()).toMatchSnapshot();
    });
}