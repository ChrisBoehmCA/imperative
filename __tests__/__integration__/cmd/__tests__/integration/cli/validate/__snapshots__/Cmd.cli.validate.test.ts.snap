// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cmd-cli should display proper messages when missing positional parameter 1`] = `
"
Syntax Error:
You must specify one of the following options for this command:
[--option-to-specify-1, --option-to-specify-2, --option-to-specify-3]

Syntax Error:
If you do not specify the following option:
--absence-implies

You must specify one of these options:
[--implied-by-absence]

Example:

 - example 1:

      $ cmd-cli validate syntax \\"position1\\" \\"testposition\\"

Use \\"cmd-cli validate syntax --help\\" to view command description, usage, and options.
"
`;

exports[`cmd-cli should display proper messages when missing positional parameter 2`] = `
"================ISSUING CMD WITH INVALID POSITIONAL ARGUMENT===============
"
`;
